core/src/main/java/org/apache/calcite/sql/SqlTableIdentifierWithID.java

```diff
@@ -0,0 +1,314 @@
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.util.SqlVisitor;
+import org.apache.calcite.sql.validate.SqlMonotonicity;
+import org.apache.calcite.sql.validate.SqlTableIdentifierWithIDQualified;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.sql.validate.SqlValidatorScope;
+import org.apache.calcite.tools.ValidationException;
+import org.apache.calcite.util.Litmus;
+import org.apache.calcite.util.Util;
+
+import com.google.common.collect.ImmutableList;
+
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.checkerframework.dataflow.qual.Pure;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * A <code>SqlTableIdentifierWithID</code> is an identifier, possibly compound.
+ * This differs from SqlIdentifier because it only refers to tables that will have
+ * an extra column not found in their type, __bodo_row_id. We use this identifier
+ * to uniquely identify particular tables of interest, for example the target table
+ * in MergeInto.
+ */
+public class SqlTableIdentifierWithID extends SqlNode {
+  /**
+   * Array of the components of this compound identifier.
+   **
+   */
+  public ImmutableList<String> names;
+
+  /**
+   * This identifier's collation (if any).
+   */
+  final @Nullable SqlCollation collation;
+
+  /**
+   * A list of the positions of the components of compound identifiers.
+   */
+  protected @Nullable ImmutableList<SqlParserPos> componentPositions;
+
+  /** Create a SqlTableIdentifierWithID from identifier components. This is used in case
+   * we make a copy. This is mostly copied from SQLIdentifier.
+   */
+  public SqlTableIdentifierWithID(
+      List<String> names,
+      @Nullable SqlCollation collation,
+      SqlParserPos pos,
+      @Nullable List<SqlParserPos> componentPositions) {
+    super(pos);
+    this.names = ImmutableList.copyOf(names);
+    this.collation = collation;
+    this.componentPositions = componentPositions == null ? null
+        : ImmutableList.copyOf(componentPositions);
+    for (String name : names) {
+      assert name != null;
+      // We don't support wildcards
+      assert !name.equals("");
```

Comment:
Isn't wildcard "*"? Or do wildcards have empty names by default?

This is likely correct, I just want clarification.

Reply:
The wildcard is represented in the list of names as the empty string. This is explained in SqlIdentifier.